### DISCUSSION TITLE
MultiSamlStrategy on NestJS

### DISCUSSION THREAD

**Discussion by dovepa** [bergie/passport-saml#873] [General]
Hello everyone,
As stated in the title, my implementation of MultiSamlStrategy on NestJS is not working.

Here's a code example:

```javascript
export class AzureStrategy extends PassportStrategy(MultiSamlStrategy, 'multi-saml') {
    constructor() {
        super({
                passReqToCallback: true, // makes req available in callback
                getSamlOptions: function (request, done) {
                    findProvider(request, function (err, provider) {
                        if (err) {
                            return done(err);
                        }
                        return done(null, provider.configuration);
                    });
                }
            },
            function (req, profile, done) {
                findByEmail(profile.email, function (err, user) {
                    if (err) {
                        return done(err);
                    }
                    return done(null, user);
                });
            })
    }
}
```

I don't understand why, but I never enter the callback function. 
Additionally, I noticed that the certificates are not loaded into the strategy, even though I have properly added them to the configuration. 
I also do not enter the **certsToCheck** function of the library.

Do you know what could be causing the problem?

Version 3.2.4

**Comment by cjbarth** [comment]
It appears you are using a deprecated version of `passport-saml`. Please upgrade to the latest version. This the note on [this](https://www.npmjs.com/package/passport-saml) page.

**Comment by dovepa** [comment]
Hello @cjbarth ,

Thank you for your answer, but I have exactly the same behavior with  "@node-saml/passport-saml": "^4.0.4",

**Reply by cjbarth** [reply]
I've changed to a discussion because there is no bug identified here yet. If there is, we can open an issue with a clear repro. I don't know why you're not entering the callback. If you do a very simple block like:

```javascript
export class AzureStrategy extends PassportStrategy(MultiSamlStrategy, 'multi-saml') {
    constructor() {
        super({
                passReqToCallback: true, // makes req available in callback
                getSamlOptions: function (request, done) {
                   console.log('getSamlOptions');
                }
            },
            function (req, profile, done) {
                console.log('callback');
            })
    }
}
```

Do either of those messages get logged?

**Reply by dovepa** [reply]
After testing, I never log _callback_, but only _getSamlOptions_.

**Reply by cjbarth** [reply]
So, for things to work, you first have to make sure that you're calling `done` with the options object you'd like. What does your `findProvider()` function look like?

**Reply by dovepa** [reply]
My `findProvider()` ~return~ use a config obj like : 
```javascript 
config : {
  issuer: 'XXXXXXXXXXXXX',
  tenantid: 'XXXXXXXXXXXXXX',
  callbackUrl: 'https://XXXXXXXXX/bop/api/login/callback',
  entryPoint: 'https://login.microsoftonline.com/XXXXXXXXXXXXX/saml2',
  signatureAlgorithm: 'sha256',
  name: 'XXXXXXXXXX',
  logoutUrl: '',
  cert: '-----BEGIN CERTIFICATE----- XXXXXXXXXX-----END CERTIFICATE-----'
}
```
Did i forget something ?

**Reply by cjbarth** [reply]
It shouldn't return anything. It should call `done()` with the config (seems like the documentation is wrong on this point).

However, as I read your code more closely, I see that you're calling the `ctor()` with too few parameters; there should be 3. Have you set breakpoints in the code to see what might be going wrong? Do you have error handling set up correctly? We have pretty good coverage of errors in the code, so if some data is wrong, it should throw.

**Reply by dovepa** [reply]
yes exactly, it doesn't return anything sorry, it uses config in the done function

**Reply by dovepa** [reply]
the parameters don't seem to be required in the constructor.
moreover authentication works well.... 
but the lib dont check certificate

**Reply by cjbarth** [reply]
The `ctor` takes 3 arguments in v4. See https://github.com/node-saml/passport-saml/blob/master/src/multiSamlStrategy.ts#L27

Are you saying that [this](https://github.com/node-saml/node-saml/blob/master/src/xml.ts#L76) code never gets called? I suggest you put a breakpoint somewhere and figure out why.

**Reply by dovepa** [reply]
Understood. I have conducted tests with three arguments inside the constructor, and the authentication is functioning correctly.
However, I am facing an issue where the certificates specified inside the configuration are not being utilized.

To explain further, there appears to be no distinction between two test cases using different certificates:
cert: '-----BEGIN CERTIFICATE----- BOB-----END CERTIFICATE-----'
cert: '-----BEGIN CERTIFICATE----- LILI-----END CERTIFICATE-----' ( certificate is invalid)
Despite using different certificate values, there is no discernible difference in behavior. 

Additionally, I am not entering the callback functions (signonVerify, logoutVerify)...

**Reply by cjbarth** [reply]
Have a look at [this](https://github.com/node-saml/node-saml/blob/58495fee0cd64148bc4dccdc57c2d701c41acbd4/test/test-signatures.spec.ts#L103) test. Note that it fails if you change the cert. When you look at the test suite you'll note that there are many things you can test without ever using the cert. I don't know how you're testing, but it may be that you're just not doing anything that requires checking the cert.

**Reply by srd90** [reply]
> I don't know how you're testing, but it may be that you're just not doing anything that requires checking the cert.

Which would mean that there is something shady about implementation of this endpoint at @dovepa's codebase:
>`https://xxxxxxxxx/bop/api/login/callback`

quoted value was taken from this comment: https://github.com/node-saml/passport-saml/discussions/873#discussioncomment-6607468

getSamlOptions is called at multiple locations/phases and one of those locations is the one that triggers autentication (and moves control to IdP) another is when UE comes back from IdP to SP via front channel  (to callback endpoint which should handle autentication request i.e. aforementioned endpoint) after IdP has autenticated end user. So you might be reporting that getSamlOptions is called during login initiator phase. More interesting thing is what happens after IdP has autenticated user. 

So… could you @dovepa show how you have implemented this https://github.com/node-saml/passport-saml/tree/v4.0.4#provide-the-authentication-callback for your `/bop/api/login/callback` (and does IdP instruct UE to move to that endpoint at all?)

**Reply by dovepa** [reply]
Thank you very much for your guidance. Indeed, the callback request had no: `@UseGuards(AuthGuard('multi-saml'))`.

If someone ever has the same problem, here's the corrected callback:

```typescript
@Post('login/callback')
@UseGuards(AuthGuard('multi-saml'))
async azurelogincallback(@Req() _request: Request, @Res() _res: any, @Body() _body) {.........}
```
